### PR TITLE
Set up spinaker pipeline

### PR DESCRIPTION
**Pull request by andreavs**
The goal of this is PR is to set up a spinaker pipeline that automatically runs the docker image on cognitedata-development whenever there is a new push to master

### CODE REVIEW

**Comment by andreavs** [comment]
@psalaberria002 This is my first time setting up a pipeline - do i need anything besides this?

File: Jenkinsfile [hunk 1]
```diff
@@ -36,6 +36,8 @@ podTemplate(
         envVar(key: 'COGNITE_BASE_URL', value: "https://greenfield.cognitedata.com"),
         envVar(key: 'COGNITE_CLIENT_NAME', value: "powerdummy-populator")
     ]) {
+
+    deploySpinnakerPipelineConfigs {}
```
Comment (psalaberria002) [inline Jenkinsfile]: Move this up to the second line of this file. It creates a separate pod when running the jenkins job, so it doesn't really belong inside this podTemplate.

File: Jenkinsfile [hunk 2]
```diff
@@ -69,7 +71,10 @@ podTemplate(
                     sh('#!/bin/sh -e\n' + 'docker login -u _json_key -p "$(cat /jenkins-docker-builder/credentials.json)" https://eu.gcr.io')
                     sh("docker push $dockerImageName:$dockerImageTag")
                 }
+                stageWithNotify('Send deployment message to Spinnaker','spinnaker-deployment-trigger') {
+                    deployToSpinnaker(dockerImageName: dockerImageName, dockerImageTag: dockerImageTag, kubernetesManifests: "manifests.yaml")
```
Comment (psalaberria002) [inline Jenkinsfile:74]: manifests.yaml needs to exist in this repo, so rename manifest.yaml to manifests.yaml :)

File: manifest.yaml [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+metadata:
+  name: powerdummy-populator
+  namespace: relationships
+  labels:
+    app: powerdummy
+    team: relationships
+spec:
+  replicas: 1
+  strategy:
+    type: Recreate
```
Comment (psalaberria002) [inline manifest.yaml]: really? Unless you really need this behavior (never more than 1 pod, zero during updates), you should leave this section with the defautl strategy (just remove it)

File: manifest.yaml [hunk 4]
```diff
@@ -0,0 +1,43 @@
+---
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: powerdummy-populator
+  namespace: relationships
+  labels:
+    app: powerdummy
```
Comment (psalaberria002) [inline manifest.yaml]: be consistent with the app label. powerdummy-populator or powerdummy ?

File: manifest.yaml [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+  strategy:
+    type: Recreate
+  revisionHistoryLimit: 5
+  selector:
+    matchLabels:
+      app: powerdummy-populator
+  template:
+    metadata:
+      annotations:
+        prometheus.io/scrape: "true"
```
Comment (psalaberria002) [inline manifest.yaml]: Delete the annotations if you are not exposing metrics

File: manifest.yaml [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+      labels:
+        app: powerdummy-populator
+        team: relationships
+    spec:
+      terminationGracePeriodSeconds: 180
```
Comment (psalaberria002) [inline manifest.yaml]: Remove this. Default is usually good.

File: manifest.yaml [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+      containers:
+        - name: powerdummy-populator
+          image: eu.gcr.io/cognitedata/powerdummy-populator
+          resources:
+            requests:
```
Comment (psalaberria002) [inline manifest.yaml]: add limits also

```
limits:
  memory: 600Mi
  cpu: 500m
```
for example

File: spinnaker-config/powerdummy-populator/powerdummy.yaml [hunk 8]
```diff
@@ -0,0 +1,21 @@
+schema: "1"
+id: powerdummyPopulator
+pipeline:
+  application: powerdummy
```
Comment (psalaberria002) [inline spinnaker-config/powerdummy-populator/powerdummy.yaml]: `powerdummy-populator`. Needs to match the subfolder name.

File: spinnaker-config/powerdummy-populator/powerdummy.yaml [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+  name: Repopulate powerdummy tenant
+  template:
+    source: spinnaker://deploySingleClusterFromPubsub
+  variables:
+    githubRepository: powerdummy-populator
+    kubernetesAccount: cognitedata-development
+    kubernetesManifests: manifests.yaml
+    dockerOrganization: cognitedata-development
```
Comment (psalaberria002) [inline spinnaker-config/powerdummy-populator/powerdummy.yaml]: delete this line. The default org is cognitedata, and the images are pushed there from jenkins.

File: spinnaker-config/powerdummy-populator/powerdummy.yaml [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+    dockerImageName: powerdummy-populator
+    autoDeployOfficeHours: false
```
Comment (psalaberria002) [inline spinnaker-config/powerdummy-populator/powerdummy.yaml]: remove this, it's only available in multistage pipelines

File: spinnaker-config/powerdummy-populator/powerdummy.yaml [hunk 11]
```diff
@@ -0,0 +1,21 @@
...
+    manualJudgement: false
```
Comment (psalaberria002) [inline spinnaker-config/powerdummy-populator/powerdummy.yaml]: Same for this one I think, you don't need it.

**Review by psalaberria002 — CHANGES_REQUESTED** [review]

**Comment by psalaberria002** [comment]
Do you really need a deployment, or a Job? Looks like what you want is a Kubernetes Job that runs once after every merge to master.

**Comment by sanderland** [comment]
> Do you really need a deployment, or a Job? Looks like what you want is a Kubernetes Job that runs once after every merge to master.

Sounds like we need a job.  
... better update my cv 

... I'll let my self out

File: Jenkinsfile (hunk 1 above)
Comment (andreavs) [reply Jenkinsfile]: fixed

File: Jenkinsfile (hunk 2 above)
Comment (andreavs) [reply Jenkinsfile:74]: oops. Fixed

File: manifest.yaml (hunk 4 above)
Comment (andreavs) [reply manifest.yaml]: fixed

File: manifest.yaml (hunk 3 above)
Comment (andreavs) [reply manifest.yaml]: fixed

File: manifest.yaml (hunk 5 above)
Comment (andreavs) [reply manifest.yaml]: fixed

File: manifest.yaml (hunk 6 above)
Comment (andreavs) [reply manifest.yaml]: fixed

File: manifest.yaml (hunk 7 above)
Comment (andreavs) [reply manifest.yaml]: fixed

File: spinnaker-config/powerdummy-populator/powerdummy.yaml (hunk 8 above)
Comment (andreavs) [reply spinnaker-config/powerdummy-populator/powerdummy.yaml]: fixed

File: spinnaker-config/powerdummy-populator/powerdummy.yaml (hunk 9 above)
Comment (andreavs) [reply spinnaker-config/powerdummy-populator/powerdummy.yaml]: fixed

File: spinnaker-config/powerdummy-populator/powerdummy.yaml (hunk 10 above)
Comment (andreavs) [reply spinnaker-config/powerdummy-populator/powerdummy.yaml]: fixed

File: spinnaker-config/powerdummy-populator/powerdummy.yaml (hunk 11 above)
Comment (andreavs) [reply spinnaker-config/powerdummy-populator/powerdummy.yaml]: fixed

**Comment by andreavs** [comment]
Yes, a job is what we need. Does that entail anything other than changing Deployment to Job in the manifest?

**Comment by psalaberria002** [comment]
The podTemplate probably stays the same, but the first part is different. We. have many jobs. Check the smoke test manifests in the api repo for example.

**Comment by andreavs** [comment]
ah, thanks. I think i got it right now. Maybe. Perhaps.

**Comment by psalaberria002** [comment]
which manifests did you use as base for the job?

**Comment by andreavs** [comment]
started from the relationships search loader, as i figured it should be similar (i.e. that just runs a docker image that starts a dataflow job) - realized now that the search loader is a bit more complex, because it has some stuff to ensure the job redeploys if it is brought down (which is why it is a deployment rather than i job, i guess)

File: manifests.yaml [hunk 12]
```diff
@@ -0,0 +1,38 @@
+---
+apiVersion: batch/v1
+kind: Job
+metadata:
+  name: powerdummy-populator
+  namespace: relationships
```
Comment (psalaberria002) [inline manifests.yaml:6]: add this

annotations:
    strategy.spinnaker.io/recreate: "true"

File: manifests.yaml [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+spec:
+  replicas: 1
+  revisionHistoryLimit: 5
+  selector:
```
Comment (psalaberria002) [inline manifests.yaml]: don't need the selector, remove it

File: manifests.yaml [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+          env:
+            - name: COGNITE_BASE_URL
+              value: "https://greenfield.cognitedata.com"
+            - name: COGNITE_CLIENT_NAME
+              value: "powerdummy-populator"
+            - name: POWERDUMMY_API_KEY
+              valueFrom:
+                secretKeyRef:
+                  name: powerdummy-apikey
+                  key: api-key
```
Comment (psalaberria002) [inline manifests.yaml:37]: add restartPolicy and backoffLimit see https://github.com/cognitedata/api/blob/master/api-smoke-test.yaml#L83

**Review by psalaberria002 — CHANGES_REQUESTED** [review]

File: manifests.yaml (hunk 12 above)
Comment (andreavs) [reply manifests.yaml:6]: fixed

File: manifests.yaml (hunk 13 above)
Comment (andreavs) [reply manifests.yaml]: fixed

File: manifests.yaml (hunk 14 above)
Comment (andreavs) [reply manifests.yaml:37]: fixed.. i assume "Never" and "0" are resonable here as well

**Comment by psalaberria002** [comment]
Did you forget to push your changes?

**Comment by andreavs** [comment]
...not my greatest night of coding

File: spinnaker-config/powerdummy-populator/app-config.yaml [hunk 15]
```diff
@@ -0,0 +1,3 @@
+#name: relationship-api
```
Comment (psalaberria002) [inline spinnaker-config/powerdummy-populator/app-config.yaml]: it's just a comment, bt still. rename this

**Review by psalaberria002 — APPROVED** [review]

**Review by sanderland — APPROVED** [review]

**Pull request merged**